File: python/tests/test_errors.py
"""Tests for error types."""

from aegis.errors import (
    AegisError,
    ApiError,
    AuthenticationError,
    NetworkError,
    NotFoundError,
    QuotaExceededError,
    RateLimitError,
    ServerError,
    TierAccessError,
    ValidationError,
)


class TestErrorHierarchy:
    def test_base_error(self):
        err = AegisError("test")
        assert str(err) == "test"
        assert err.message == "test"

    def test_api_error(self):
        err = ApiError("bad request", status_code=400, error_code="INVALID", request_id="req-1")
        assert err.status_code == 400
        assert err.error_code == "INVALID"
        assert err.request_id == "req-1"
        assert "[400]" in str(err)
        assert "(INVALID)" in str(err)
        assert "req-1" in str(err)

    def test_authentication_error(self):
        err = AuthenticationError()
        assert err.status_code == 401
        assert isinstance(err, ApiError)
        assert isinstance(err, AegisError)

    def test_tier_access_error(self):
        err = TierAccessError(
            "Upgrade required",
            required_tier="pro",
            upgrade_url="https://aiaegis.io/upgrade",
        )
        assert err.status_code == 403
        assert err.required_tier == "pro"
        assert err.upgrade_url == "https://aiaegis.io/upgrade"

    def test_quota_exceeded_error(self):
        err = QuotaExceededError(limit=1000, used=1001, reset_at="2026-03-01T00:00:00Z")
        assert err.status_code == 429
        assert err.limit == 1000
        assert err.used == 1001
        assert err.reset_at == "2026-03-01T00:00:00Z"

    def test_rate_limit_error(self):
        err = RateLimitError(retry_after=60.0)
        assert err.status_code == 429
        assert err.retry_after == 60.0

    def test_validation_error(self):
        err = ValidationError("Invalid prompt", status_code=422)
        assert err.status_code == 422

    def test_not_found_error(self):
        err = NotFoundError()
        assert err.status_code == 404

    def test_server_error(self):
        err = ServerError(status_code=503)
        assert err.status_code == 503

    def test_network_error(self):
        cause = ConnectionError("refused")
        err = NetworkError("Connection failed", cause=cause)
        assert err.cause is cause
        assert isinstance(err, AegisError)

File: rust/tests/test_errors.rs
use aegis_sdk::AegisError;

#[test]
fn error_display_implementations() {
    let auth = AegisError::Authentication {
        message: "Invalid key".into(),
        request_id: Some("req-1".into()),
    };
    let s = auth.to_string();
    assert!(!s.is_empty());
    assert!(s.contains("Authentication") || s.contains("Invalid key"));

    let tier = AegisError::TierAccess {
        message: "Upgrade required".into(),
        required_tier: Some("pro".into()),
        upgrade_url: Some("https://aiaegis.io/upgrade".into()),
    };
    let s = tier.to_string();
    assert!(!s.is_empty());
    assert!(s.contains("Tier") || s.contains("Upgrade"));

    let quota = AegisError::QuotaExceeded {
        message: "Quota exceeded".into(),
        limit: Some(1000),
        used: Some(1001),
        reset_at: Some("2025-02-25T00:00:00Z".into()),
    };
    let s = quota.to_string();
    assert!(!s.is_empty());
    assert!(s.contains("Quota") || s.contains("exceeded"));

    let rate_limit = AegisError::RateLimit {
        message: "Rate limited".into(),
        retry_after: Some(60.0),
    };
    let s = rate_limit.to_string();
    assert!(!s.is_empty());
    assert!(s.contains("Rate") || s.contains("limited"));

    let validation = AegisError::Validation {
        message: "Validation failed".into(),
        status_code: 422,
    };
    let s = validation.to_string();
    assert!(!s.is_empty());
    assert!(s.contains("Validation") || s.contains("failed"));

    let not_found = AegisError::NotFound {
        message: "Not found".into(),
    };
    let s = not_found.to_string();
    assert!(!s.is_empty());
    assert!(s.contains("Not found"));

    let server = AegisError::Server {
        message: "Server error".into(),
        status_code: 500,
    };
    let s = server.to_string();
    assert!(!s.is_empty());
    assert!(s.contains("Server") || s.contains("error"));
}

#[test]
fn error_variants_match_expected_messages() {
    let auth = AegisError::Authentication {
        message: "Invalid API key".into(),
        request_id: None,
    };
    assert!(auth.to_string().contains("Invalid API key"));

    let tier = AegisError::TierAccess {
        message: "Access denied".into(),
        required_tier: Some("enterprise".into()),
        upgrade_url: Some("https://aiaegis.io/plans".into()),
    };
    let s = tier.to_string();
    assert!(s.contains("Access denied") || s.contains("enterprise") || s.contains("plans"));

    let quota = AegisError::QuotaExceeded {
        message: "Limit reached".into(),
        limit: Some(100),
        used: Some(100),
        reset_at: Some("2025-02-25T00:00:00Z".into()),
    };
    assert!(quota.to_string().contains("Limit reached") || quota.to_string().contains("Quota"));
}
